### PR TITLE
release/darwin: Create universal binary

### PR DESCRIPTION
**Pull request by NickeZ**
Create a binary that supports aarch64 (Apple M1)

### CODE REVIEW

**Comment by NickeZ** [comment]
@x1ddos I fixed the `lipo` call and created a PR. Please test it on your M1 system.

File: bitbox-bridge/release/Dockerfile
Comment (x1ddos) [inline bitbox-bridge/release/Dockerfile:39]: I couldn't get past this line: "Unable to locate package llvm-11" and clang-11.

image

I'm actually surprised the whole thing always requires docker image and osxcross. I built the universal binary without docker and no issues, running these commands manually:

```shell
$ rustup default stable-x86_64-apple-darwin
$ cargo build --target x86_64-apple-darwin --release
$ file target/x86_64-apple-darwin/release/bitbox-bridge
target/x86_64-apple-darwin/release/bitbox-bridge: Mach-O 64-bit executable x86_64

$ rustup default stable-aarch64-apple-darwin
$ cargo build --target aarch64-apple-darwin --release
$ file target/aarch64-apple-darwin/release/bitbox-bridge
target/aarch64-apple-darwin/release/bitbox-bridge: Mach-O 64-bit executable arm64

$ lipo -create -output bitbox-bridge-uni bitbox-bridge-x86 bitbox-bridge-aarch64
$ file bitbox-bridge-uni 
bitbox-bridge-uni: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64]
bitbox-bridge-uni (for architecture x86_64):    Mach-O 64-bit executable x86_64
bitbox-bridge-uni (for architecture arm64):    Mach-O 64-bit executable arm64
```

If I don't `rustup default stable-x86_64-apple-darwin` and just do `cargo build --target x86_64-apple-darwin --release` while the default is aarch64, I get this eror:

```
$ cargo build --target x86_64-apple-darwin --release
   Compiling bitbox-bridge v1.3.0 (/Users/dev/code/bitbox-bridge/bitbox-bridge)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-arch" "x86_64" "-L" "/Users/dev/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/dev/code/bitbox-bridge/target/x86_64-apple-darwin/release/deps/bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o" "-o" "/Users/dev/code/bitbox-bridge/target/x86_64-apple-darwin/release/deps/bitbox_bridge" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/dev/code/bitbox-bridge/target/x86_64-apple-darwin/release/deps" "-L" "/Users/dev/code/bitbox-bridge/target/release/deps" "-L" "/Users/dev/code/bitbox-bridge/target/x86_64-apple-darwin/release/build/hidapi-8b1a006045e6977b/out" "-L" "/Users/dev/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/var/folders/vy/k8bwyxjn3gl82q9xnflthzxc0000gp/T/rustcE9bEMc/libhidapi-f5e0d83556ec0aba.rlib" "/Users/dev/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-10db70d883838cbc.rlib" "-framework" "Security" "-framework" "IOKit" "-framework" "CoreFoundation" "-lSystem" "-lresolv" "-lc" "-lm"
  = note: ld: warning: ignoring file /var/folders/vy/k8bwyxjn3gl82q9xnflthzxc0000gp/T/rustcE9bEMc/libhidapi-f5e0d83556ec0aba.rlib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
          Undefined symbols for architecture x86_64:
            "_hid_enumerate", referenced from:
                hidapi::HidApi::get_hid_device_info_vector::h8b637a262beebcc1 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_free_enumeration", referenced from:
                hidapi::HidApi::get_hid_device_info_vector::h8b637a262beebcc1 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_error", referenced from:
                hidapi::HidDevice::check_error::h4cfdf7834a51e571 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_close", referenced from:
                alloc::sync::Arc$LT$T$GT$::drop_slow::h137692c626eb9049 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
                core::ptr::drop_in_place$LT$hidapi..HidDevice$GT$::h240b36821f117494 (.11655) in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_read", referenced from:
                std::sys_common::backtrace::__rust_begin_short_backtrace::hedf06b6242c8bdad in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_write", referenced from:
                _$LT$$RF$mut$u20$T$u20$as$u20$futures_io..if_std..AsyncWrite$GT$::poll_write::h74e007385314096e in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_init", referenced from:
                bitbox_bridge::main::hfb7940008c45e7c9 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_exit", referenced from:
                bitbox_bridge::main::hfb7940008c45e7c9 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
                alloc::sync::Arc$LT$T$GT$::drop_slow::h793e7207a23e0f23 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
                core::ptr::drop_in_place$LT$hidapi..HidApiLock$GT$::ha3f65a2732b78843 in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
            "_hid_open_path", referenced from:
                _$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$::poll::h247d2b4eafd7e85f in bitbox_bridge.bitbox_bridge.6ojjinjr-cgu.0.rcgu.o
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I suspect somewhere in the middle aarch64 is used (bug?) and hid lib is built for wrong architecture. But this is a minor issue imho. Maybe setting some additional rustc flags would fix it, but setting the default arch before building the whole thing also works.

Oh and I also ` mv .cargo/config .cargo/config.org` before starting to compile anything, just in case.

Can we get rid of osxcross?

**Review by x1ddos — COMMENTED** [review]
Couldn't test with the docker image - see the other comment - but the rest of the changes work :+1:

File: bitbox-bridge/release/darwin/release.sh
Comment (NickeZ) [reply bitbox-bridge/release/darwin/release.sh:22]: I run this in docker on Linux, so for me it is needed :/

File: bitbox-bridge/release/Dockerfile
Comment (NickeZ) [reply bitbox-bridge/release/Dockerfile:39]: This build script is for Linux, so we need it, one option is to have build instructions for osx as well which should be simpler

**Comment by NickeZ** [comment]
Generally the idea was to build everything in docker with specified versions of everything and then package in osx and Windows but I'm open to improving the release flow of you have suggestions.

File: bitbox-bridge/release/darwin/release.sh
Comment (NickeZ) [reply bitbox-bridge/release/darwin/release.sh:17]: Yeah, I also realized that and removed them in the other open PR

File: bitbox-bridge/release/darwin/package.sh
Comment (NickeZ) [reply bitbox-bridge/release/darwin/package.sh:13]: Comment could be improved, but the scripts should be executed from the root project dir. They use paths relative to the working dir which won't work otherwise

**Comment by x1ddos** [comment]
Well, my reasoning is: you'll need an apple hardware/os to release for 
macos due to codesign and notarization. You'll need windows to sign the 
the final binary.

So, given that you'll need Windows and macOS anyway, the purpose of the 
"universal" docker image which builds for all platforms is suddently not 
as useful. Might as well build directly on that platform anyway.

The advantage is obvious to me: no need to even have docker installed on 
on those platforms. I had actually some issues with installing docker on 
M1 and eventually gave up and uninstalled it.

Not sure how it works on Windows but I don't really have high hopes.

**Comment by NickeZ** [comment]
The intent was never to use docker on osx or Windows. Docker was used to set up a reproducible cross build environment on Linux.

**Comment by NickeZ** [comment]
I would need to build in VMs if I would build natively and that takes significantly more time as well. Not all free CIs have osx and Windows machines also :/ maybe github does, I'm not up to date

**Comment by x1ddos** [comment]
You say the idea is to build everything on linux and then package on the respective platforms but you see https://github.com/digitalbitbox/bitbox-bridge/pull/25 now says to run `cargo install toml-echo` and at that point you'd already have rust installed on macOS. Once you have rust installed, it's literally 3 commands: (1) build arm, (2) build amd64 and (3) lipo to compose universal binary.

So, at least for macOS, building on linux doesn't really have any advantages imho. Anyway, I'll try a final build as is in a linux container and then we'll see how to simplify.

**Comment by benma** [comment]
> So, at least for macOS, building on linux doesn't really have any advantages imho.

Will need to at least add a rust-toolchain file to lock in the version.

**Comment by NickeZ** [comment]
> You say the idea is to build everything on linux and then package on the respective platforms but you see #25 now says to run `cargo install toml-echo` and at that point you'd already have rust installed on macOS. Once you have rust installed, it's literally 3 commands: (1) build arm, (2) build amd64 and (3) lipo to compose universal binary.

Yes I assume that everyone has the rust compiler installed. Even mac users.

> So, at least for macOS, building on linux doesn't really have any advantages imho. Anyway, I'll try a final build as is in a linux container and then we'll see how to simplify.

I build more often than I package/run, so to keep the iteration time short I prefer to cross compile. PR welcome to add osx scripts, but please don't remove the scripts for cross compiling since that would slow me down. (yes I could have two vms open or wait for CI to build on all platforms, but why make something fast slow?)

**Comment by NickeZ** [comment]
> How often does it happen that linux builds but macos or windows doesn't?

I don't know

> And assuming they all built with a success, how do you know it actually runs?

With a type system like rusts, if it compiles it works ;)

**Comment by x1ddos** [comment]
> > How often does it happen that linux builds but macos or windows doesn't?
> 
> I don't know
> 
> > And assuming they all built with a success, how do you know it actually runs?
> 
> With a type system like rusts, if it compiles it works ;)

Then why do you even bother cross-compiling? :)

**Review by x1ddos — APPROVED** [review]
LGTM after updating the llvm comment in Dockerfile.

**Comment by x1ddos** [comment]
@NickeZ forgot to add: could you also update the change log to mention that we're now producing universal binary for macOS.

**Pull request merged**
